homeassistant/components/osoenergy/water_heater.py

```diff
@@ -0,0 +1,194 @@
+            "powersave": "power_save",
+            "extraenergy": "extra_energy",
+            "voltage": "voltage",
+            "ffr": "ffr",
+        },
+    },
+    "optimization_mode": {
+        "ha_name": "optimization_mode",
+        "value_mapping": {
+            "off": STATE_OFF,
+            "oso": "oso",
+            "gridcompany": "grid_company",
+            "smartcompany": "smart_company",
+            "advanced": "advanced",
+        },
+    },
+    "profile": {"ha_name": "profile", "is_profile": True},
+    "volume": {"ha_name": "volume"},
+    "v40_min": {"ha_name": "v40_min"},
+    "v40_level_min": {"ha_name": "v40_level_min"},
+    "v40_level_max": {"ha_name": "v40_level_max"},
+}
+HEATER_MIN_TEMP = 10
+HEATER_MAX_TEMP = 80
+MANUFACTURER = "OSO Energy"
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up OSO Energy heater based on a config entry."""
+    osoenergy = hass.data[DOMAIN][entry.entry_id]
+    devices = osoenergy.session.device_list.get("water_heater")
+    entities = []
+    if devices:
+        for dev in devices:
+            entities.append(OSOEnergyWaterHeater(osoenergy, dev))
+    async_add_entities(entities, True)
+
+
+def _get_local_hour(utc_hour: int):
+    """Get the local hour."""
+    now = dt_util.utcnow()
+    now_local = dt_util.now()
+    utc_time = now.replace(hour=utc_hour, minute=0, second=0, microsecond=0)
+    local_hour = dt_util.as_local(utc_time)
+    local_hour = local_hour.replace(
+        year=now_local.year, month=now_local.month, day=now_local.day
+    )
+    return local_hour
+
+
+def _convert_profile_to_local(values):
+    """Convert UTC profile to local."""
+    profile = [None] * 24
+    for hour in range(24):
+        local_hour = _get_local_hour(hour)
+        local_hour_string = local_hour.strftime("%Y-%m-%dT%H:%M:%S%z")
+        profile[local_hour.hour] = {local_hour_string: values[hour]}
+
+    return profile
+
+
+class OSOEnergyWaterHeater(OSOEnergyEntity, WaterHeaterEntity):
+    """OSO Energy Water Heater Device."""
+
+    _attr_name = None
+    _attr_supported_features = WaterHeaterEntityFeature.TARGET_TEMPERATURE
+    _attr_temperature_unit = UnitOfTemperature.CELSIUS
+    _attr_translation_key = "saga_heater"
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return device information."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, self.device["device_id"])},
+            manufacturer=MANUFACTURER,
+            model=self.device["device_type"],
+            name=self.device["device_name"],
+        )
+
+    @property
+    def available(self) -> bool:
+        """Return if the device is available."""
+        return self.device.get("attributes", {}).get("available", False)
+
+    @property
+    def current_operation(self) -> str:
+        """Return current operation."""
+        return self.device["status"]["current_operation"]
```

Comment:
This needs to be converted to one of the allowed states: https://developers.home-assistant.io/docs/core/entity/water-heater#states